Add DeleteParser, DeleteCommand

ngys117 opened the pull request:
Add delete task and module functionality.

ngys117 committed:
Add DeleteParser

Add DeleteCommand

Currently only supports deleting tasks

Update ModHappyParser

ngys117 committed:
Merge branch 'branch-PR-review' into branch-A-Delete

# Conflicts:
#	src/main/java/seedu/duke/parsers/ModHappyParser.java

ngys117 committed:
Add DeleteParser, DeleteCommand

Modify ModuleList, TaskList

Add delete task and module functionality with these changes.

ngys117 committed:
Merge branch 'master' into branch-A-Delete

ngys117 commented:
Closes #10, closes #35, closes #49

chooyikai commented on src/main/java/seedu/duke/tasks/TaskList.java:
```diff
@@ -29,6 +31,17 @@ public Task addTask(Task t) {
+     * @param index The task number to be removed.
+     */
+    public void removeTask(int index) throws NoSuchTaskException {
+        if (index >= list.size()) {
```
Maybe check for `index < 0 || index >= list.size()` instead? Just in case.

chooyikai commented on src/main/java/seedu/duke/tasks/ModuleList.java:
```diff
@@ -21,6 +21,14 @@ public Module addModule(Module m) {
+     * @param moduleCode the module code to be removed
+     */
+    public void removeModule(String moduleCode) {
+        list.remove(getModule(moduleCode));
```
Maybe check whether `getModule()` returns `null` before attempting to call `list.remove()` on it.

Also, see above remark on returning the removed module instead.

chooyikai commented on src/main/java/seedu/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,58 @@
+            taskList.removeTask(taskIndex);
+            System.out.println(task + " has been deleted.");
+        } catch (NoSuchTaskException e) {
+            System.out.println("Failed to delete.");
```
Instead of catching this exception using try-catch and using `System.out.println()`, maybe we should use throws instead to pass it to the Main class, which will handle it for us

ngys117 committed:
Update DeleteCommand.

Add some JavaDoc.

Remove some magic literals.

chooyikai commented on src/main/java/seedu/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,58 @@
+
+    public void deleteModule(ModuleList moduleList) {
+        moduleList.removeModule(moduleCode);
+        result = moduleCode + " has been deleted.";
```
This part could be simplified if `removeModule` returned the module that was deleted instead of being a `void` method. Then you could just do `result = moduleList.removeModule(moduleCode)` + " has been deleted".

Also, maybe convert the string constant into a format string (`"%s has been deleted."`) and call String.format here instead.

chooyikai commented on src/main/java/seedu/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,58 @@
+        try {
+            Task task = taskList.getTask(taskIndex);
+            taskList.removeTask(taskIndex);
+            System.out.println(task + " has been deleted.");
```
Same as above

chooyikai commented on src/main/java/seedu/duke/tasks/TaskList.java:
```diff
@@ -29,6 +31,17 @@ public Task addTask(Task t) {
+     * Removes the specified task from the task list.
+     * @param index The task number to be removed.
+     */
+    public void removeTask(int index) throws NoSuchTaskException {
```
See above remark on returning the removed Task instead.

ngys117 committed:
Update Delete Parser

Update Delete Command

ngys117 committed:
Add JUnit Testing

Update JavaDoc

ngys117 committed:
Add NoSuchModuleException

Add exception check for non-existent module for deleteModule

Update exception check for deleteTask

chooyikai commented on src/main/java/seedu/duke/parsers/DeleteParser.java:
```diff
@@ -0,0 +1,36 @@
+            return new DeleteCommand(moduleCode);
+        }
+        if (!taskNumberString.isBlank()) {
+            return new DeleteCommand(Integer.parseInt(taskNumberString));
```
Consider wrapping `Integer.parseInt()` in a try-catch block, just in case (although this should not fail during normal code execution because the regex already filters out invalid inputs or negative numbers).

Once done it should be good to merge.

ngys117 committed:
Add exception check for converting string to int

ngys117 committed:
Shift exception check.

Previous location causes exception to always be thrown when trying to
delete a module.
